include/lbann/optimizers/data_type_optimizer_impl.hpp

```diff
@@ -93,7 +91,35 @@ auto data_type_optimizer<TensorDataType>::get_weights() const
 }
 
 template <typename TensorDataType>
-auto data_type_optimizer<TensorDataType>::get_gradient() -> AbsDistMatrixType&
+auto data_type_optimizer<TensorDataType>::get_gradient()
+  -> const AbsDistMatrixType&
+{
+  auto& sharded_gradient = this->get_gradient_sharded();
+
+  auto matrix_dist = std::get<2>(this->get_matrix_info());
+  // If the gradient is not sharded, return a view
+  if (sharded_gradient.DistData() == matrix_dist) {
+    return sharded_gradient;
+  }
+
+  // Create a new matrix with the correct value distribution (usually STAR_STAR)
+  // and copy the values from there.
+  auto result = AbsDistMatrixType::Instantiate(
+    *matrix_dist.grid,
+    matrix_dist.root,
+    matrix_dist.colDist,
+    matrix_dist.rowDist,
+    (matrix_dist.blockHeight == 1 && matrix_dist.blockWidth == 1 ? El::ELEMENT
+                                                                 : El::BLOCK),
+    matrix_dist.device);
+
+  El::Copy(sharded_gradient, *result);
+  return *result;
```

Comment:
I understand and agree with this comment. Returning a const reference was done to avoid creating an even larger pull request. I want to return an `AbstractDistMatrix<T>` (i.e., no reference). Would that incur any additional copies and would it be freed appropriately?

Reply:
You cannot return an `AbstractDistMatrix<T>` by value -- it's an abstract class. Your best bet is to return `std::unique_ptr<AbstractDistMatrix<T>>`.

Reply:
Fixed. Thanks.